File: medium/2364_count_number_of_bad_pairs/solution.cpp
#include <unordered_map>
#include <vector>

class Solution {
public:
  long long countBadPairs(std::vector<int> &nums) {
    const size_t n = nums.size();

    std::unordered_map<int, int> pairsMap;
    for (size_t i{}; i < n; ++i)
      ++pairsMap[i - nums[i]];

    long long goodPairsCount = 0;

    for (const auto &[diff, count] : pairsMap)
      goodPairsCount += static_cast<long long>(count) * (count - 1) / 2;

    return (n * (n - 1) / 2) - goodPairsCount;
  }
};

File: medium/2364_count_number_of_bad_pairs/solution.py
from collections import defaultdict


class Solution:
    def countBadPairs(self, nums: list[int]) -> int:
        n = len(nums)

        pairsMap = defaultdict(int)

        pairsCount = n * (n - 1) // 2
        goodPairsCount = 0

        for i in range(n):
            goodPairsCount += pairsMap[i - nums[i]]
            pairsMap[i - nums[i]] += 1

        return pairsCount - goodPairsCount
